examples/mnist/assets/algo_cnn/algo.py

```diff
@@ -0,0 +1,91 @@
+import substratools as tools
+import os
+import keras
+from keras import backend as K
+from keras.models import Sequential
+from keras.layers import Dense, Dropout, Flatten
+from keras.layers import Conv2D, MaxPooling2D
+
+# input image dimensions
+img_rows, img_cols = 28, 28
+input_shape = (img_rows, img_cols, 1)
+
+
+class Algo(tools.algo.Algo):
+    def _normalize_X(self, X):
+
+        if K.image_data_format() == "channels_first":
+            X = X.reshape(X.shape[0], 1, img_rows, img_cols)
+
+            input_shape = (1, img_rows, img_cols)
+        else:
+            X = X.reshape(X.shape[0], img_rows, img_cols, 1)
+
+            input_shape = (img_rows, img_cols, 1)
+
+        X = X.astype("float32")
+
+        X /= 255
+
+        print(f"X shape: {X.shape}")
+        print(f"{X.shape[0]} samples")
+
+        return X
+
+    def train(self, X, y, models, rank):
+        X = self._normalize_X(X)
+        num_classes = 10
+        y = keras.utils.to_categorical(y, num_classes)
+
+        model = Sequential()
+        model.add(
+            Conv2D(32, kernel_size=(3, 3), activation="relu", input_shape=input_shape)
+        )
+        model.add(Conv2D(64, (3, 3), activation="relu"))
+        model.add(MaxPooling2D(pool_size=(2, 2)))
+        model.add(Dropout(0.25))
+        model.add(Flatten())
+        model.add(Dense(128, activation="relu"))
+        model.add(Dropout(0.5))
+        model.add(Dense(num_classes, activation="softmax"))
+
+        model.compile(
+            loss=keras.losses.categorical_crossentropy,
+            optimizer=keras.optimizers.Adadelta(),
+            metrics=["accuracy"],
+        )
```

Comment:
You could also put that in a dedicated `initialize_new_model()` function to keep your `train()` function shorter / easier to read